rotkehlchen/chain/gnosis/transactions.py

```diff
@@ -59,32 +61,57 @@ def get_chain_specific_multiaddress_data(
         else:
             from_block = self.evm_inquirer.get_blocknumber_by_time(from_ts)
 
-        events = self.evm_inquirer.get_logs(
-            contract_address=BLOCKREWARDS_ADDRESS,
-            abi=ADDED_RECEIVER_ABI,
-            event_name='AddedReceiver',
-            # For multiple addresses unfortunately have to query all and filter later
-            argument_filters={'receiver': addresses[0]} if len(addresses) == 1 else {},
-            from_block=from_block,
-            to_block='latest',
-            call_order=self.evm_inquirer.default_call_order(),
-        )
-
-        expected_topics = ['0x000000000000000000000000' + x.lower()[2:] for x in addresses]
-        for event in events:
-            if event['topics'][1] in expected_topics:
-                with suppress(AlreadyExists):
-                    self.add_transaction_by_hash(
-                        tx_hash=deserialize_evm_tx_hash(event['transactionHash']),
-                        associated_address=hex_or_bytes_to_address(event['topics'][1]),
-                        must_exist=True,
-                    )
-
-        with self.database.user_write() as write_cursor:
-            for address in addresses:  # updated DB cache
-                self.database.update_used_query_range(
-                    write_cursor=write_cursor,
-                    name=f'{BRIDGE_QUERIED_ADDRESS_PREFIX}{address}',
-                    start_ts=from_ts,
-                    end_ts=to_ts,
-                )
+        last_block_processed = None
+        expected_topics = {'0x000000000000000000000000' + x.lower()[2:] for x in addresses}
+
+        def _process_events_cb(
+                last_block_queried: int,
+                filters: dict[str, Any],  # pylint: disable=unused-argument
+                new_events: list[dict[str, Any]],
+        ) -> None:
+            """Callback that processes new events
+            This function also keeps track of the last block number processed using
+            the variable nonlocal variable last_block_processed. It is later used
+            in case of error to save the queried range.
+            """
+            nonlocal last_block_processed
```

Comment:
no this is ugly and super error prone. What you made is a closure, but I don't think it's required in this case, plus capturing the surrounding scope with nonlocal is less performant. Why implement a closure when you can just add add optional arguments to the callback function?

It's a bit more performatn and less ugly

I do notice you modify the last_block_processed so you can also do that with optional argument but it would need to be an object you modify I suppsoe.